File: sistemi lineari/determinante_laplace.m
function det=laplace(A)

% calcola il determinante di una matrice (quadrata)
%  mediante la regola di laplace applicata nella prima riga

% Parametri di input : 
% A: matrice quadrata

% Parametri di output : 
% det: determinante di A

[m,n]=size(A);

if m~=n 
    error('matrice non quadrata')
end

if n==1
    det=A;
else
    det=0;
    for j=1:n

        % metodo abbreviato
        A1j=A(2:m,[1:j-1,j+1:n]);

        % metodo corretto ma con più righe
        % A1j = A;
        % A1j(1,:) = [];
        % A1j(:,j) = [];

        det=det+(-1)^(1+j)*A(1,j)*laplace(A1j);
    end
end

File: sistemi lineari/determinante_laplace.py
from numpy import * 

def determinante_laplace(A):
    
    """
    Regola di Laplace per il calcolo del determinante di una matrice quadrata.
    Solleva errore se la matrice non è quadrata.
    
    INPUT:
        A : matrice quadrata di cui calcolare il determinante
    OUTPUT:
        d : determinante della matrice A
    """
    
    m,n = shape(A)
    
    if m != n:
        raise("La matrice non è quadrata")
    
    
    # Fissata la 1° riga
    
    if n==1:
        d = A[0,0]  # ritorniamo il valore semplice dell'elemento
    else:
        d = 0
        for j in range(0,n):
            A1j = delete(delete(A,0,axis=0),j,axis=1)   # A1j è la sottomatrice di A ottenuta scartando la prima riga e la j-esima colonna
            d = d +(-1)**(j)*A[0,j]*determinante_laplace(A1j)   
            
    return d

# esempio
A = array([ [1,0,-2], [-1,0,1], [1,1,-3]])
print("A : \n",A)
print("\nDeterminante di A : ",determinante_laplace(A))
